shared/js/passcode_helper.js

```diff
@@ -0,0 +1,161 @@
+/* exported PasscodeHelper */
+/* globals crypto, TextEncoder */
+(function(exports) {
+  'use strict';
+
+  const SET_DIGEST_VALUE = 'lockscreen.passcode-lock.digest.value';
+  const SET_DIGEST_SALT = 'lockscreen.passcode-lock.digest.salt';
+  const SET_DIGEST_ITERATIONS = 'lockscreen.passcode-lock.digest.iterations';
+  const SET_DIGEST_ALGORITHM = 'lockscreen.passcode-lock.digest.algorithm';
+  const DEFAULT_ALGORITHM = 'SHA-1'; //XXX Update to SHA-256 after bug 554827?
+  const DEFAULT_ITERATIONS = 1000;
+
+
+  var PasscodeHelper = function() {
+    /* Usage:
+     *      var ph = new PasscodeHelper();
+     *      ph.setPassccode(string) -> Promise => digest (or false)
+     *      ph.checkPasscode(string) -> Promise resolves to a boolean
+     * */
+
+
+    var _salt;
+    var _iterations;
+    var _algorithm;
+    var _testPass;
+
+    this._encode = function(str) {
+      return new TextEncoder('utf-8').encode(str);
+    };
+    this._toTypedArray = function(obj) {
+      // SettingsAPI doesnt like arrays and gives us an Object { 1: .., 2: .. }
+      var a = [];
+      for (var key in obj) {
+        a.push(obj[key]);
+      }
+      return new Uint8Array(a);
+    };
+    this._deriveBits = function(pwKey, salt, iterations, algorithm) {
+      var length = 256;
```

Comment:
Nevertheless, if it would make code more complicated than what we want here, we can temporary adopt tight coupling like this. However if you have further plan for substitute the derivation method, we should discuss to adopt other patterns like Strategy patter to re-organize our code.